### DISCUSSION TITLE
🚀  Arm-hosted Runners - Public Beta Feedback

### DISCUSSION THREAD

**Discussion by grandmas-favorite** [community/community#127102] [Actions] [answered]
Hi Community!

We're excited to unveil ArmⓇ-based Linux and Windows runners for GitHub Actions, **available today (June 3rd, 2024) to our customers on our GitHub Team and Enterprise Cloud plan**!

⚙️ This new addition to our suite of hosted runners provides power, performance and sustainability improvements for all your GitHub Actions jobs. Developers can now take advantage of Arm-based hardware hosted by GitHub to build and deploy their release assets anywhere Arm architecture is used. Best of all, these runners are priced at 37% less than our x64 Linux and Windows runners.

### Documentation and Resources
To help you get started, we have prepared comprehensive documentation and resources:

*   🗞️    **GitHub Blog**: Read about the full release on the GitHub change log, [here](https://github.blog/2024-06-03-arm64-on-github-actions-powering-faster-more-efficient-build-systems/).

*   📊  **Docs and Monitoring**: Find out more information on how to set up Arm-hosted runners in [this video](https://www.youtube.com/watch?v=vrr_OgMk458) or in our public [documentation](https://docs.github.com/actions/using-github-hosted-runners/about-larger-runners/about-larger-runners). To learn more about runner-per-minute pricing, check out the [documentation](https://docs.github.com/billing/managing-billing-for-github-actions/about-billing-for-github-actions#per-minute-rates).

*   📝   **More Help**: To view the list of installed software, give feedback, or report issues with the image, head to the new [partner runner images repository](https://github.com/actions/partner-runner-images).

### Participate in the Public Beta!
Your feedback during this beta phase is invaluable to us. We encourage you to share your experiences here, which will be instrumental in refining and enhancing this feature as we look toward the future.

Stay tuned for updates and enhancements throughout the beta period!

**Comment by theofficialgman** [comment]
Can't wait for the availability for all GitHub users by the end of the year https://github.com/github/roadmap/issues/970

Can you confirm the exact ARM processors being used? According to the roadmap the Ampere Altra family of chips are being used https://github.com/github/roadmap/issues/970.

I am particularly interested in whether the ARM processors used will always be capable of executing ARM 32bit instructions natively (EL0 AArch32). ~~Almost all (currently manufactured) ARMv8.X-A processors can execute 32bit and 64bit instructions (except Apple's cores, A65, A65AE, A34, and Neoverse V1) while all ARMv9.X-A processors can only execute 64bit instructions (32bit instruction support was dropped in the silicon design).~~

The Ampere Altra family of chips can execute 32bit (EL0 AArch32) and 64bit instructions (https://amperecomputing.com/solutions/arm-native).

Will there ever be a time where the AmpereOne family of processors will be used? The AmpereOne family of processors does not have this defined so its uncertain to me if they support 32bit instructions or not. I would not like if some runners use AmpereOne at some point in the future and fail EL0 AArch32 and some use Ampere Altra and succeed.

I am aware that all runner images will be ARM64, I am not asking for runner images for ARMhf incase there was any confusion there.

**Reply by theofficialgman** [reply]
For any users, if you have trouble checking for that information, you can test if AArch32 EL0 works by running an appropriate binary

as long as no qemu binfmt is installed, this should succeed
```
wget https://github.com/Azathothas/Static-Binaries/raw/main/wget/wget_busybox_armv7l_abihf_musl_Linux
chmod +x wget wget_busybox_armv7l_abihf_musl_Linux
./wget_busybox_armv7l_abihf_musl_Linux --help
```

and this should fail
```
wget https://github.com/Azathothas/Static-Binaries/raw/main/wget/wget_busybox_amd_x86_64_musl_Linux
chmod +x wget_busybox_amd_x86_64_musl_Linux
./wget_busybox_amd_x86_64_musl_Linux --help
```

if both succeed then there is a QEMU binfmt allowing x86_64 instructions to be emulated. remove all binfmts to make sure that the armv7l still succeeds and x86_64 fails.

**Reply by lkfortuna** [reply]
Hey @theofficialgman  - I can confirm that the runenrs in the public beta are currently using the Ampere Altra chip. 

As far as future specifications, we will most likely also have the cobalt chips from Azure in the future as well. I cannot say for sure whether we will ever have AmpereOne though.

**Reply by theofficialgman** [reply]
@lkfortuna Microsoft Azure Cobalt 100 CPU's use ARM Neoverse N2 cores, an ARMv9.0-A core ~~which (as is the case with all ARMv9 cores) lacks AArch32 EL0 support. I hope that you make these Cobalt chips either available under a separate runner name or able to be disabled through some yaml option. If some runners have Ampere Altra chips and others have Azure Cobalt 100 chips this will cause my actions (and I am sure many others who intend to use the AArch32 EL0 support) to fail unexpectedly (when the action gets deployed to a Azure Cobalt 100 chip).~~

edit: upon further investigation it appears that AArch32 EL0 support is present on Arm Neoverse N2 cores https://developer.arm.com/documentation/102099/0000/The-Neoverse-N2--core/Neoverse-N2--core-features?_ga=2.57483618.912127831.1635187957-327063191.1554242723 . Further documentation about the AArch32/AArch64 exception levels https://developer.arm.com/documentation/102412/0103/Execution-and-Security-states/Impact-of-implemented-Exception-levels

**Reply by theofficialgman** [reply]
To summarize, AAarch32 EL0 support is an optional feature of ARMv8 and ARMv9. The majority of ARMv8 implementations include it while its yet to be seen if most ARMv9 cores will.

ARM Designed cores that include AArch32 EL0 support:
Neoverse N1, N2, V1

cores that do NOT:
Neoverse E1, V2, V3, V3AE, N3

Ampere Altra is a Neoverse N1 based design and Microsoft Azure Cobalt 100 is a Neoverse N2 based design so both support AArch32 EL0.

I will note that the Neoverse N3 and V3 are the latest cores and both lack AArch32 EL0 support. This is concerning for the future but not today.

**Comment by EsterKais** [comment]
It seems the announcement is missing a link to the video 😊 

>Find out more information on how to set up Arm-hosted runners in **this** video or ..

I assume you wanted to link this: https://www.youtube.com/watch?v=vrr_OgMk458? 😊

**Reply by grandmas-favorite** [reply]
Nice catch @EsterKais!

Post is updated with the video link 😁

**Reply by albertocavalcante** [reply]
Is the video now hosted elsewhere?

![image](https://github.com/user-attachments/assets/47fbde0f-8f76-4d00-84e8-ab0c60184890)

**Reply by lkfortuna** [reply]
Yes, is is here! https://www.youtube.com/watch?v=cgI6SBP8pEM

**Comment by chojnicki** [comment]
I was waiting for this, but it's annoying that I cannot use custom runner on non-organization project. I bought Team plan for that, but to be able to use created runner I'm forced to move my repositories ownership to my organization - even that my organization is just me.. I cannot do that because it will also change URL of repository which I cannot do for many reasons. I hope ARM will be available for all (not just open source like it states in docs/news), but still it will be not possible to use larger runner anyway. 

I WANT to pay for custom runners, but github/microsoft does not want my money. Instead I'm forced to buy servers elsewhere and host own runners. It's strange business strategy.

**Reply by lkfortuna** [reply]
hi @chojnicki  - thank you for the feedback! I can confirm that when we ship these runners for opensource, they will be available for personal namespaces as well which should solve the repository issue for you.

**Reply by theofficialgman** [reply]
I have the same issues presented here. I am happy to pay for a Pro account (same cost as Team for one user) but according to the announcement it isn't available for such plans.

**Comment by NikhilKumar7777** [comment]
It seems like you're looking for information or guidance related to Arm-hosted runners in the context of a public beta feedback. This is a relatively new feature that GitHub and other CI/CD platforms have started to offer, allowing users to run their workflows on Arm architecture machines.

Here are some steps and points to consider when working with Arm-hosted runners and providing feedback:

Setting Up Arm-hosted Runners
Registering the Runner: You need to register an Arm-hosted runner in your GitHub repository or organization. You can follow GitHub's official documentation to do this.

Configuration: Configure your GitHub Actions workflow to use the Arm-hosted runner. You might need to specify the runner type in your workflow YAML file.

Example workflow YAML:

name: CI

on: [push]

jobs:
  build:
    runs-on: [self-hosted, ARM]

    steps:
    - uses: actions/checkout@v2
    - name: Run a one-line script
      run: echo Hello, world!
    - name: Run a multi-line script
      run: |
        echo Add other necessary steps here
        echo e.g., installing dependencies, running tests, etc.


**Summary:**
A brief summary of your feedback, including the overall experience with Arm-hosted runners.

**Performance:**
- Detailed performance metrics (e.g., build times, execution times)
- Comparison with x86 runners

**Compatibility:**
- List of compatible and incompatible dependencies/tools
- Any workarounds used

**Resource Utilization:**
- Observations on CPU, memory usage, etc.

**Stability:**
- Any stability issues encountered
- Frequency and nature of failures

**Documentation and Support:**
- Evaluation of existing documentation
- Areas where documentation could be improved

**Feature Requests:**
- Suggested features or improvements

**General Impressions:**
- Overall impressions and experiences
- Highlights of positive or negative aspects

**Reply by queenofcorgis** [reply]
@NikhilKumar7777 
We’ve clarified our stance on using generative AI tools like ChatGPT within our Community via [this announcement](https://github.com/orgs/community/discussions/69006). Please review the guidelines to ensure your post meets them as failure to adhere to those rules can result in action taken by our moderator team.  You can read our updated [Code of Conduct](https://github.com/community/community/blob/baa5cd9020ef3f2bbf86dca1191d9c90f9b0822d/CODE_OF_CONDUCT.md#reasonable-use-of-ai-generated-content) and the [announcement](https://github.com/orgs/community/discussions/69006) for more details. Thank you for helping us maintain an authentic and beneficial space for everyone.

**Comment by ubidefeo** [comment]
Link to the video leads to an unavailable content page

**Comment by chris-griffin** [comment]
I took a stab at changing from the default Linux x64 runners using `ubuntu-20.04` to what I presumed was an equivalent Linux arm64 using "Base Image for Ubuntu Server 20.04". I ran into a handful of issues which I note below.

**1. No easy way to identify what the installed software will be for the 3 partner images available**

I quickly realized that the images available for Linux arm64 do not contain the same installed software as the GitHub support `ubuntu-20.04` image for Linux x64. That is not very surprising as GitHub does denote these as "Partner" images.

<img width="949" alt="image" src="https://github.com/community/community/assets/1795659/bfc45085-db75-4656-afe9-0c759884eaeb">

However, the [Learn more about partner images](https://github.com/actions/partner-runner-images) link in the above screenshot only provides details on the [ARM ubuntu 22 image](https://github.com/actions/partner-runner-images/blob/main/images/arm-ubuntu-22-image.md). Information on the other two images is not available.

Similarly, there is no link available in the "Set up job" step to the installed software
<img width="960" alt="image" src="https://github.com/community/community/assets/1795659/67527499-dd35-465d-9a4c-fcde2fde03df">
<img width="726" alt="image" src="https://github.com/community/community/assets/1795659/05fc1053-5a29-4120-a140-8e7a2e098b98">


**2. `actions/setup-python` does not work for the latest version of python 3.8, 3.9 or 3.10 despite those being supported**

While the `actions/setup-python` and `actions/python-versions` repos did recently add support for Linux arm64, many of the latest supported python patch versions are inextricably missing. I opened https://github.com/actions/setup-python/issues/892 with more details.

**Reply by lkfortuna** [reply]
Hey @chris-griffin thank you for the feedback! 
1- You're right, there is no documentation on the base images. To answer, these are just base image with no tools pre-installed on them, they jsut provide the OS framework. We will deprecate these images eventually in favor of the ones with tooling. The only arm64 image with tools pre-installed on it currently is the "ubuntu 22 by arm limited" image. Thank you for calling out this confusion, will try and clear that up in our documentation. 
2- For setup-python support this should be coming on monday, but for the missing patches you opened an issue in the right place!

**Comment by crohr** [comment]
Great to see some action on that front! I've just added the GitHub ARM runners to the benchmark of GitHub Actions providers: https://runs-on.com/benchmarks/github-actions-runners/#arm64-runners. 

Regularly updated, to see the various performance in terms of CPU speed, queuing time, etc.

**Comment by sandstrom** [comment]
The new ARM runners are rate-limited on Docker hub:

https://github.com/actions/runner-images/issues/1445#issuecomment-2208581589

More background:

https://github.com/actions/runner-images/issues/1445#issuecomment-713861495

Are there any plans to solve this rate-limiting issue?

**Reply by lkfortuna** [reply]
hi there, yes! We are working on fixing this, will update here when that is done.

**Reply by sandstrom** [reply]
@lkfortuna Thanks Larissa!

**Reply by lkfortuna** [reply]
@sandstrom - sorry for delay in getting back to you, but this should be fixed as of a couple weeks ago.

**Comment by samus-aran** [comment, marked as answer]
## Arm-hosted Runners Now Generally Available!

Check out our [Changelog blog post](https://github.blog/changelog/2024-09-03-github-actions-arm64-linux-and-windows-runners-are-now-generally-available/) for more details 👈

**Reply by natescherer** [reply]
Is there any update on when arm64 runners will be available for open-source projects? [This blogpost](https://github.blog/news-insights/product-news/arm64-on-github-actions-powering-faster-more-efficient-build-systems/) says "We expect to begin offering Arm runners for open source projects by the end of the year." (meaning 2024) in the last paragraph, but, as far as I can tell, they are not yet available.

**Reply by ghost** [reply]
It was moved to "Q1 2025 – Jan-Mar" - [ARM hosted runners for OSS/free plans [Public Preview] #970](https://github.com/github/roadmap/issues/970).

<img width="311" alt="Screenshot 2025-01-03 at 13 04 40" src="https://github.com/user-attachments/assets/162d7496-4a4d-4717-85f3-36eb30c4bce5" />

**Reply by natescherer** [reply]
Thanks!

**Reply by lkfortuna** [reply]
we shipped linux arm64 runners for free today: https://github.blog/changelog/2025-01-16-linux-arm64-hosted-runners-now-available-for-free-in-public-repositories-public-preview/
